jscomp/others/js_array.ml

```diff
@@ -17,61 +17,122 @@
+(** should we use [bool] or [boolan] seems they are intechangeable here *)
+external filter : ('a -> bool [@bs]) -> 'this = "" [@@bs.send.pipe: 'a t as 'this]
+external filteri : ('a -> int  -> Js.boolean[@bs]) -> 'this = "filter" [@@bs.send.pipe: 'a t as 'this]
+
+external find : ('a -> bool [@bs]) -> 'a Js.undefined = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external findi : ('a -> int -> bool [@bs]) -> 'a Js.undefined  = "find" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
 
+external findIndex : ('a -> bool [@bs]) -> int = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external findIndexi : ('a -> int -> bool [@bs]) -> int = "findIndex" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
 
-external forEach : ('a -> unit [@bs]) -> unit  = "" [@@bs.send.pipe: 'a t as 'this]
+external forEach : ('a -> unit [@bs]) -> unit = "" [@@bs.send.pipe: 'a t as 'this]
 external forEachi : ('a -> int -> unit [@bs]) -> unit  = "forEach" [@@bs.send.pipe: 'a t as 'this]
 
-external map : ('a  ->  'b [@bs]) -> 'b t  = "" [@@bs.send.pipe: 'a t as 'this]
-external mapi : ('a -> int ->  'b [@bs]) -> 'b t = "map" [@@bs.send.pipe: 'a t as 'this]
+external keys : int array_iter = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
 
-(** should we use [bool] or [boolan] seems they are intechangeable here *)
-external filter : ('a  -> bool [@bs]) -> 'this = "" [@@bs.send.pipe: 'a t as 'this]
-external filteri : ('a -> int  -> Js.boolean[@bs]) -> 'this = "filter" [@@bs.send.pipe: 'a t as 'this]
+external map : ('a  -> 'b [@bs]) -> 'b t = "" [@@bs.send.pipe: 'a t as 'this]
+external mapi : ('a -> int ->  'b [@bs]) -> 'b t = "map" [@@bs.send.pipe: 'a t as 'this]
 
-external reducei : ('a -> 'a -> int -> 'a [@bs]) ->  'a -> 'a = "reduce" [@@bs.send.pipe: 'a t as 'this]
-external reduce :  ('a -> 'a  -> 'a [@bs]) ->  'a -> 'a = "reduce" [@@bs.send.pipe: 'a t as 'this]
+external reduce :  ('b -> 'a  -> 'b [@bs]) -> 'b -> 'b = "" [@@bs.send.pipe: 'a t as 'this]
+external reducei : ('b -> 'a -> int -> 'b [@bs]) -> 'b -> 'b = "reduce" [@@bs.send.pipe: 'a t as 'this]
 
-external isArray : 'a -> Js.boolean = "Array.isArray" [@@bs.val]
+external reduceRight :  ('b -> 'a  -> 'b [@bs]) -> 'b -> 'b = "" [@@bs.send.pipe: 'a t as 'this]
+external reduceRighti : ('b -> 'a -> int -> 'b [@bs]) -> 'b -> 'b = "reduceRight" [@@bs.send.pipe: 'a t as 'this]
 
-(** https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push *)    
-external push : 'a array  -> 'a -> int = ""
-[@@bs.send]
+external some : ('a  -> Js.boolean [@bs]) -> Js.boolean = "" [@@bs.send.pipe: 'a t as 'this]
+external somei : ('a  -> int -> Js.boolean [@bs]) -> Js.boolean = "some" [@@bs.send.pipe: 'a t as 'this]
 
-external pop : 'a array -> 'a Js.undefined = ""
-[@@bs.send]
-   
-external length : 'a array -> int = ""
-[@@bs.get]
-   
+external values : 'a array_iter = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
```

Comment:
ah, so this is ES6 iterator, maybe we should not expose here, we need do some research to see how we could support iterator pattern in Buckle